### PR TITLE
Typing indicator (typing chat bubble)

### LINKED ISSUE #4915: Add typing indicators
Game needs typing indicators, y'know, the little text bubble that appears on top of your head when you're typing, lets other people know you're typing, duh. I'm sure this is bikeshed material, since time ago I saw people talking about people shooting while you're typing.

### PR DESCRIPTION
**Pull request by Macoron**
## About the PR 
Fix #4915
Continues #5173

**Screenshots**

https://user-images.githubusercontent.com/6161335/168661075-db1f52f3-eeca-4b92-bfbc-36417e008f8b.mp4

**Changelog**

:cl:
- add: Added chat typing indicator.

### CODE REVIEW

**Review by mirrorcult — CHANGES_REQUESTED** [review]
looks good, a couple requests:

- people have expressed desire for a client cvar that makes it so they don't show typing indicators, doesnt seem like that would be too hard to do?
- the solution of making it a layer on the entity that gets added and the data is synced with appearance data is good to me, but want a second opinion
- how does it look for mobs stacked on the same tile (or pAIs on the same tile as someone else)?

File: Content.Server/Chat/TypingIndicator/TypingIndicatorSystem.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        SubscribeNetworkEvent<TypingChangedEvent>(OnClientTypingChanged);
+    }
+
+    private void OnPlayerAttached(PlayerAttachedEvent ev)
+    {
+        // when player poses entity we want to make sure that there is typing indicator
+        EnsureComp<TypingIndicatorComponent>(ev.Entity);
+        // we also need appearance component to sync visual state
+        EnsureComp<ServerAppearanceComponent>(ev.Entity);
+    }
```
[the remark this answers is not in the document]
Comment (Macoron) [reply Content.Server/Chat/TypingIndicator/TypingIndicatorSystem.cs:28]:
```
    /// <summary>
    ///     Component required for entities to be able to speak. (TODO: Entities can speak fine without this, this only forbids them speak if they have it and enabled is false.)
    ///     Contains the option to let entities make noise when speaking, datafields for the sounds in question, and relevant AudioParams.
    /// </summary>
    [RegisterComponent]
    public sealed class SharedSpeechComponent : Component
```

I'm not sure about it. Check TODO comment. Also see typing crowbars and other ghost roles is kinda cool.

**Comment by Macoron** [comment]
> people have expressed desire for a client cvar that makes it so they don't send typing indicators, doesnt seem like that would be too hard to do?

Done. In future PR this will need UI element in chat window, but for now you can disable it on a client by `cvar chat.show_typing_indicator false`. You will still see others typing indicators though.

> how does it look for mobs stacked on the same tile (or pAIs on the same tile as someone else)?

Sprite sorting looks good when entities really close to each other (just tested it). If pAIs or mouse placed inside mobs inventory, typing indicator won't be shown.

File: Content.Server/Chat/TypingIndicator/TypingIndicatorSystem.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mirrorcult) [reply Content.Server/Chat/TypingIndicator/TypingIndicatorSystem.cs:28]: sure, thats fine for now then

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**